### PR TITLE
batch2 management operations api

### PR DESCRIPTION
**Pull request by fil512**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java [hunk 1]
```diff
@@ -111,6 +113,21 @@ public Optional<JobInstance> fetchInstanceAndMarkInProgress(String theInstanceId
 		return fetchInstance(theInstanceId);
 	}
 
+	public List<JobInstance> fetchInstancesByJobDefinitionIdAndStatus(String theJobDefinitionId, Set<StatusEnum> theRequestedStatuses, int thePageSize, int thePageIndex) {
```
Comment (TipzCM) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:116]: oh man, if only i had gotten this earlier 😁 

Ah well...

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IBatch2JobInstanceRepository.java [hunk 2]
```diff
@@ -35,9 +39,15 @@ public interface IBatch2JobInstanceRepository extends JpaRepository<Batch2JobIns
...
 	@Modifying
 	@Query("UPDATE Batch2JobInstanceEntity e SET e.myCancelled = :cancelled WHERE e.myId = :id")
-	void updateInstanceCancelled(@Param("id") String theInstanceId, @Param("cancelled") boolean theCancelled);
+	int updateInstanceCancelled(@Param("id") String theInstanceId, @Param("cancelled") boolean theCancelled);
 
 	@Modifying
 	@Query("UPDATE Batch2JobInstanceEntity e SET e.myCurrentGatedStepId = :currentGatedStepId WHERE e.myId = :id")
 	void updateInstanceCurrentGatedStepId(@Param("id") String theInstanceId, @Param("currentGatedStepId") String theCurrentGatedStepId);
+
+	@Query("SELECT e FROM Batch2JobInstanceEntity e WHERE e.myDefinitionId = :jobDefinitionId AND e.myStatus IN :statuses")
```
Comment (TipzCM) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IBatch2JobInstanceRepository.java:48]: One of the questions i had about this, generally, is i wonder if we want to add the status to the index, or make a new index with the status?

I guess it depends how common such requests are going to be (i have no idea how many JobInstances an "average" database will have)

**Review by TipzCM — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:116]: Why didn't you add it yourself?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IBatch2JobInstanceRepository.java (hunk 2 above)
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IBatch2JobInstanceRepository.java:48]: We clean up the job table pretty regularly.  I think this is a case where a full table scan is fine.

**Pull request merged**